src/include/repository.coffee

```diff
@@ -496,17 +496,23 @@ module.exports=
     name: 'punctuation.separator.comma.haskell'
     match: /,/
   lit_num: [
+    name: 'constant.numeric.hexfloat.haskell'
+    match: '0[xX]_*[0-9a-fA-F](_*[0-9a-fA-F])*((\\.[0-9a-fA-F](_*[0-9a-fA-F])*(_*[pP][+-]?[0-9](_*[0-9])*)*)|(_*[pP][+-]?[0-9](_*[0-9])*))'
+  ,
     name: 'constant.numeric.hexadecimal.haskell'
-    match: '0[xX][0-9a-fA-F]+'
+    match: '0[xX]_*[0-9a-fA-F](_*[0-9a-fA-F])*'
```

Comment:
I'm not entirely sure why you're using this rather convoluted expression here (and other places). For all intents and purposes, `0[xX]_*[0-9a-fA-F](_*[0-9a-fA-F])*` is the same as `0[xX](?:_*[0-9a-fA-F])+`. And if we're not aiming for technical correctness (which we aren't), latter expression is roughly equivalent to `0[xX][_0-9a-fA-F]+`, which would also accept cases like `0x_` (which I don't think is valid Haskell), but is measurably faster.

Also please don't introduce new capturing groups if you can avoid it, it can potentially break some stuff, and test coverage isn't 100%.

Reply:
Thank you for telling me your policy.

> I'm not entirely sure why you're using this rather convoluted expression here (and other places)

I simply implemented it according to the accepted formal specification.
  * https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0009-numeric-underscores.rst#new-syntax-this-proposal

> ... but is measurably faster.

You mean that this project has adopted a policy that gives priority to speed over accuracy.
I understand.
It is one of the valuable decisions for many platforms.

> Also please don't introduce new capturing groups if you can avoid it, it can potentially break some stuff, and test coverage isn't 100%.

OK. Is it also better to merge the binary with octal?

In consideration of speed, I propose the simplified two plans below.
Are you acceptable for this?

Plan A: `BinaryLiterals` + `NumericUnderscores` extension (without `HexFloatLiterals`)

```diff
   lit_num: [
     name: 'constant.numeric.hexadecimal.haskell'
-    match: '0[xX][0-9a-fA-F]+'
+    match: '0[xX][_0-9a-fA-F]+'
   ,
     name: 'constant.numeric.octal.haskell'
-    match: '0[oO][0-7]+'
+    match: '0[oObB][_0-7]+'
   ,
     name: 'constant.numeric.float.haskell'
-    match: '[0-9]+(\\.[0-9]+[eE][+-]?|\\.|[eE][+-]?)[0-9]+'
+    match: '[_0-9]+(\\.[_0-9]+[eE][+-]?|\\.|[eE][+-]?)[_0-9]+'
   ,
     name: 'constant.numeric.decimal.haskell'
-    match: '[0-9]+'
+    match: '[0-9][_0-9]*'
   ]
```

Plan B: Plan A + `HexFloatLiterals`

This is difference for Plan A.

```diff
     name: 'constant.numeric.float.haskell'
-    match: '[_0-9]+(\\.[_0-9]+[eE][+-]?|\\.|[eE][+-]?)[_0-9]+'
+    match: '(?:0[xX])?[_0-9a-fA-F]+(?:\\.[_0-9a-fA-F]+[eEpP][+-]?|\\.|[eEpP][+-]?)[_0-9a-fA-F]+'
```

Thank you for your time.

Reply:
> Is it also better to merge the binary with octal?

No, those are different beasts, and parser has a rather obvious choice based on 2-character lookahead.

Same goes for hexadecimal floats, those are obviously (from point of view of parser) syntactically different from decimal floats. There's no need to create horrifyingly convoluted regular expressions (also apparently incorrect, `(?:0[xX])?[_0-9a-fA-F]+(?:\\.[_0-9a-fA-F]+[eEpP][+-]?|\\.|[eEpP][+-]?)[_0-9a-fA-F]+` matches `0xfade1`, which I don't think it should). Just abstract away a common pattern for floats. F.ex.:

```coffee
floatPattern = (digit, exp) -> "#{digit}+(\\.#{digit}+#{exp}[+-]?|\\.|#{exp}[+-]?)#{digit}+"
```

and then use it for both regular floats and hexfloats:
```coffee
float: "[0-9]#{floatPattern('[0-9_]', '[eE]')}" # see below why [0-9] at the start
hexfloat: "0[xX]#{floatPattern('[0-9a-fA-F_]','[pP]')}"
```

Everything else seems reasonable. Except maybe in 'Plan A'  I would use the same approach in `float` as in `decimal` (only for the first digit mind)
```regex
[0-9][_0-9]*(\\.[_0-9]+[eE][+-]?|\\.|[eE][+-]?)[_0-9]+
```
so that it could bail early on holes (like `f _x _y z = z`)

Reply:
> No, those are different beasts, and parser has a rather obvious choice based on 2-character lookahead.

Thank you. I understand.
I was misunderstanding the matter of the capturing groups.


> Just abstract away a common pattern for floats. F.ex.:

Beautiful!
I realized what you pointed out.


> Everything else seems reasonable. Except maybe in 'Plan A' I would use the same approach in float as in decimal

OK, the policy of pattern matching is below.
* The leading of the pattern should be accurate for identification with other tokens.
* Subsequent patterns prioritize speed.

I will push the revised version after this.